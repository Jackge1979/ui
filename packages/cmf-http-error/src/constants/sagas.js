// used at the last line of sagas unit tests to be sure that the saga ends and does not hang
export const SAGA_FINISHED = { done: true, value: undefined };
export const SAGA_FINISHED_TRUE = { ...SAGA_FINISHED, value: true };
export const SAGA_FINISHED_FALSE = { ...SAGA_FINISHED, value: false };
