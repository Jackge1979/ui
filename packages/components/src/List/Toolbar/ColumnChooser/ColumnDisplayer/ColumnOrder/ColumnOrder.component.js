import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ActionButton from '../../../../../Actions/ActionButton';
import theme from '../ColumnDisplayer.scss';

export const OrderDisplay = ({ order, length }) => (
	<React.Fragment>
		<span
			className={classNames(
				theme['tc-column-displayer-order-value'],
				'tc-column-displayer-order-value',
			)}
		>
			{order}
		</span>
		{`/ ${length}`}
	</React.Fragment>
);

OrderDisplay.propTypes = {
	order: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
	length: PropTypes.number.isRequired,
};

function isOrderCorrect(value, length) {
	if (Number.isInteger(value) && (value <= length && value > 0)) {
		return true;
	}
	throw new Error(`ColumnDisplayer, isOrderCorrect : Bad order number = ${value}`);
}

const ColumnOrder = ({ length, locked, value, t, ...rest }) => {
	const [editMode, setEditMode] = useState(false);
	const [ctrlValue, setCtrlValue] = useState(value);
	useEffect(() => {
		setCtrlValue(value);
	}, [value]);

	function onBlur(event) {
		const formatValue = parseInt(ctrlValue, 10);
		if (formatValue !== value) {
			if (isOrderCorrect(formatValue, length)) {
				rest.onBlur(event, formatValue);
			}
		}
		setEditMode(prevState => !prevState);
	}

	function onKeyPress(event) {
		const formatValue = parseInt(ctrlValue, 10);
		if (event.key === 'Enter') {
			if (isOrderCorrect(formatValue, length)) {
				rest.onKeyPress(event, formatValue);
				setEditMode(prevState => !prevState);
			}
		}
	}

	if (locked || !editMode) {
		return (
			<ActionButton
				aria-label={t('INPUT_TEXT_EDIT_COLUMN_CHOOSER', { defaultValue: 'Edit order' })}
				disabled={locked}
				link
				onClick={() => setEditMode(!editMode)}
				label={<OrderDisplay order={ctrlValue} length={length} />}
			/>
		);
	}
	return (
		<React.Fragment>
			<input
				autoFocus
				aria-label={t('INPUT_TEXT_ORDER_COLUMN_CHOOSER', {
					defaultValue: 'Input order',
				})}
				className={classNames(
					theme['tc-column-displayer-order-input-text'],
					'tc-column-displayer-order-input-text',
				)}
				onBlur={event => onBlur(event)}
				onChange={event => setCtrlValue(event.target.value)}
				onKeyPress={event => onKeyPress(event)}
				placeholder={ctrlValue}
				type="text"
				value={ctrlValue}
			/>
			{`/ ${length}`}
		</React.Fragment>
	);
};

ColumnOrder.propTypes = {
	length: PropTypes.number.isRequired,
	locked: PropTypes.bool,
	onBlur: PropTypes.func.isRequired,
	onKeyPress: PropTypes.func.isRequired,
	value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
	t: PropTypes.func.isRequired,
};

export default ColumnOrder;
