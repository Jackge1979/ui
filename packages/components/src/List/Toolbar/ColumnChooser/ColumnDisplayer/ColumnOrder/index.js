import Component from './ColumnOrder.component';

export default Component;
