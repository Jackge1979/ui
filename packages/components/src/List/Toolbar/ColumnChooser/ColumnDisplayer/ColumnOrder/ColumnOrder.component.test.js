import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import ColumnOrder, { OrderDisplay } from './ColumnOrder.component';
import { WSAVERNOTSUPPORTED } from 'constants';

const t = jest.fn((_, translationValue) => translationValue.defaultValue);

describe('ColumnOrder', () => {
	it('should render in non edit mode', () => {
		// given
		const props = {
			length: 3,
			order: 1,
			locked: true,
			value: 1,
			onBlur: jest.fn(),
			onKeyPress: jest.fn(),
			t,
		};
		// when
		const wrapper = mount(<ColumnOrder {...props} />);
		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});
	it('should render in edit mode', () => {
		// given
		const props = {
			length: 3,
			order: 1,
			locked: false,
			value: 1,
			onBlur: jest.fn(),
			onKeyPress: jest.fn(),
			forceEditMode: true,
			t,
		};
		// when
		const wrapper = mount(<ColumnOrder {...props} />);
		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});
});

describe('ColoumnOrder event and hook', () => {
	it('should change the current value', () => {
		// given
		const props = {
			length: 3,
			order: 1,
			locked: false,
			value: 1,
			onBlur: jest.fn(),
			onKeyPress: jest.fn(),
			t,
		};
		// when
		const wrapper = mount(<ColumnOrder {...props} />);
		expect(wrapper.find('button').text()).toEqual('1/ 3');
		wrapper.find('button').simulate('click');
		wrapper.find('input').simulate('change', { target: { value: 15 } });
		// then
		expect(wrapper.find('input').prop('value')).toBe(15);
	});
});

describe('OrderDisplay', () => {
	it('should return', () => {
		// given
		const props = {
			order: 1,
			length: 10,
		};
		// when
		const wrapper = shallow(<OrderDisplay {...props} />);
		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});
});
