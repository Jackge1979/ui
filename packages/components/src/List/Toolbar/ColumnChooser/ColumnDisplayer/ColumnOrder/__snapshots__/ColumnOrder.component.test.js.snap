// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnOrder should render in edit mode 1`] = `
<ColumnOrder
  forceEditMode={true}
  length={3}
  locked={false}
  onBlur={[MockFunction]}
  onKeyPress={[MockFunction]}
  order={1}
  t={
    [MockFunction] {
      "calls": Array [
        Array [
          "INPUT_TEXT_EDIT_COLUMN_CHOOSER",
          Object {
            "defaultValue": "Edit order",
          },
        ],
        Array [
          "INPUT_TEXT_EDIT_COLUMN_CHOOSER",
          Object {
            "defaultValue": "Edit order",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": "Edit order",
        },
        Object {
          "isThrow": false,
          "value": "Edit order",
        },
      ],
    }
  }
  value={1}
/>
`;

exports[`ColumnOrder should render in non edit mode 1`] = `
<ColumnOrder
  length={3}
  locked={true}
  onBlur={[MockFunction]}
  onKeyPress={[MockFunction]}
  order={1}
  t={
    [MockFunction] {
      "calls": Array [
        Array [
          "INPUT_TEXT_EDIT_COLUMN_CHOOSER",
          Object {
            "defaultValue": "Edit order",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": "Edit order",
        },
      ],
    }
  }
  value={1}
/>
`;

exports[`OrderDisplay should return 1`] = `
<React.Fragment>
  <span
    className="theme-tc-column-displayer-order-value tc-column-displayer-order-value"
  >
    1
  </span>
  / 10
</React.Fragment>
`;
